### PR TITLE
[feat] Support tp mode for DeepSeek-R1-W4AFP8

### PR DESCRIPTION
**Pull request by chenxijun1029**
## Motivation

Support tp mode for DeepSeek w4a8 model, which has a better performace than ep mode.

## Modifications

1. Add W4AFp8MoEMethod and associated `create_weights`, `process_weights_after_loading` function and `apply` function. In the apply function, we use the same cutlass_w4a8_moe kernel as ep moe uses.
2. Add some tile shape and cluster shape config for tp moe in `cutlass_w4a8_moe` kernel.
3. Add a router logic in w4afp8 quant config and method. When "enable_ep_moe" found in global_server_args_dict, we use ep mode, else tp.

Co-author: @yuhyao <827623970@qq.com>

## Benchmark
We run DeepSeek-R1-W4AFP8 on 8*H20 with tp8, comparing to run DeepSeek-R1 on 8*H20 with ep8.
Test configuration: 

### ISL1000, OSL1000
input/output len = 1000/1000, qps=128, max_concurrency=128, num_prompt=256.
The results are shown below:

TP
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    128.0     
Max request concurrency:                 128       
Successful requests:                     256       
Benchmark duration (s):                  159.00    
Total input tokens:                      256000    
Total generated tokens:                  256000    
Total generated tokens (retokenized):    254696    
Request throughput (req/s):              1.61      
Input token throughput (tok/s):          1610.09   
Output token throughput (tok/s):         1610.09   
Total token throughput (tok/s):          3220.18   
Concurrency:                             127.52    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   79201.47  
Median E2E Latency (ms):                 78956.21  
---------------Time to First Token----------------
Mean TTFT (ms):                          6547.98   
Median TTFT (ms):                        6612.11   
P99 TTFT (ms):                           11687.82  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           72.73     
Median ITL (ms):                         68.05     
P95 ITL (ms):                            72.42     
P99 ITL (ms):                            73.05     
Max ITL (ms):                            11148.65  
==================================================
```

While EP:
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    128.0     
Max request concurrency:                 128       
Successful requests:                     256       
Benchmark duration (s):                  161.37    
Total input tokens:                      256000    
Total generated tokens:                  256000    
Total generated tokens (retokenized):    255343    
Request throughput (req/s):              1.59      
Input token throughput (tok/s):          1586.41   
Output token throughput (tok/s):         1586.41   
Total token throughput (tok/s):          3172.81   
Concurrency:                             127.53    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   80390.85  
Median E2E Latency (ms):                 80644.43  
---------------Time to First Token----------------
Mean TTFT (ms):                          8143.41   
Median TTFT (ms):                        8144.56   
P99 TTFT (ms):                           14833.05  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           72.32     
Median ITL (ms):                         66.38     
P95 ITL (ms):                            71.43     
P99 ITL (ms):                            72.08     
Max ITL (ms):                            14113.65  
==================================================
```

### ISL6000, OSL1000
input/output len = 6000/1000, qps=128, max_concurrency=128, num_prompt=256.
The results are shown below:

TP
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    128.0     
Max request concurrency:                 128       
Successful requests:                     256       
Benchmark duration (s):                  525.30    
Total input tokens:                      1536000   
Total generated tokens:                  256000    
Total generated tokens (retokenized):    254578    
Request throughput (req/s):              0.49      
Input token throughput (tok/s):          2924.03   
Output token throughput (tok/s):         487.34    
Total token throughput (tok/s):          3411.37   
Concurrency:                             109.09    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   223849.73 
Median E2E Latency (ms):                 251328.94 
---------------Time to First Token----------------
Mean TTFT (ms):                          118601.26 
Median TTFT (ms):                        149225.11 
P99 TTFT (ms):                           183862.13 
---------------Inter-Token Latency----------------
Mean ITL (ms):                           105.36    
Median ITL (ms):                         83.10     
P95 ITL (ms):                            85.79     
P99 ITL (ms):                            86.27     
Max ITL (ms):                            52034.42  
==================================================
```

While EP:
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    128.0     
Max request concurrency:                 128       
Successful requests:                     256       
Benchmark duration (s):                  592.58    
Total input tokens:                      1536000   
Total generated tokens:                  256000    
Total generated tokens (retokenized):    255781    
Request throughput (req/s):              0.43      
Input token throughput (tok/s):          2592.07   
Output token throughput (tok/s):         432.01    
Total token throughput (tok/s):          3024.08   
Concurrency:                             108.41    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   250942.46 
Median E2E Latency (ms):                 282080.60 
---------------Time to First Token----------------
Mean TTFT (ms):                          137965.32 
Median TTFT (ms):                        171540.14 
P99 TTFT (ms):                           217855.50 
---------------Inter-Token Latency----------------
Mean ITL (ms):                           113.09    
Median ITL (ms):                         81.36     
P95 ITL (ms):                            84.83     
P99 ITL (ms):                            85.55     
Max ITL (ms):                            71910.62  
==================================================
```

We conclude the results as the below sheet:
| Scenario                     | Version             | ITL (ms)        | TTFT (ms)            | Request throughput (req/s)        |
|--------------------------|------------------|------------------|----------------------|--------------------|
| 1K input，1K output，128 request rate | EP   | 72.32            | 8 143.41             | 1.59               |
|                          | TP   | 72.73 (+0.6 %)   | 6 547.98 (−19.60 %)  | 1.61 (+1.2 %)      |
| 6K input，1K output，128 request rate | EP   | 113.09           | 137 965.32           | 0.43               |
|                          | TP   | 105.36 (−6.83 %) | 118 601.26 (−14.03 %)| 0.49 (+14.0 %)     |

#Accuracy
mmlu: 86.9  （by @yuhyao ）
aime24: 80.0
gpqa: 71.2
math500: 95.6

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests] - The unittest for `cutlass_w4a8_moe` kernel already exsited (https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by AniZpZ** [comment]
Thanks for your great work on this! To help us evaluate the impact of this PR, could you please provide the performance results (like GSM8K, MMLU, and Hellaswag)

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 1]
```diff
@@ -40,8 +40,9 @@
     sglang_per_token_group_quant_fp8,
     sglang_per_token_quant_fp8,
 )
+from sglang.srt.layers.quantization.fp8_utils import normalize_e4m3fn_to_e4m3fnuz
```
Comment (yangsijia-celina) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: seems this import isn't used

File: python/sglang/srt/layers/quantization/w4afp8.py [hunk 2]
```diff
@@ -223,33 +253,187 @@ def create_weights(
...
+                hidden_size,
+                intermediate_size_per_partition // group_size,
+                dtype=params_dtype,
+            ),
+            requires_grad=False,
+        )
+        layer.register_parameter("w2_weight_scale_inv", w2_scales)
+        set_weight_attrs(w2_scales, extra_weight_attrs)
+
+        # The input scale for w1 and w3 should be the same
```
Comment (yangsijia-celina) [inline python/sglang/srt/layers/quantization/w4afp8.py]: just want to confirm if you've checked the contents of the act_scales.safetensors file. Are the input scales for w1 and w3 all consistent?

File: python/sglang/srt/layers/quantization/w4afp8.py [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
 import logging
-from typing import Any, Dict, List, Optional
+from typing import Any, Callable, Dict, List, Optional
 
 import torch
 from torch.nn import Module
 from torch.nn.parameter import Parameter
 
+
+from python.sglang.srt.layers.moe.cutlass_w4a8_moe import cutlass_w4a8_moe
```
Comment (huangzl18883) [inline python/sglang/srt/layers/quantization/w4afp8.py]: remove python prefix?

File: python/sglang/srt/layers/quantization/w4afp8.py [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
 import logging
-from typing import Any, Dict, List, Optional
+from typing import Any, Callable, Dict, List, Optional
 
 import torch
 from torch.nn import Module
 from torch.nn.parameter import Parameter
 
+
+from sglang.srt.layers.moe.cutlass_w4a8_moe import cutlass_w4a8_moe
```
Comment (donpromax) [inline python/sglang/srt/layers/quantization/w4afp8.py]: Python module and function cannot use the same name, otherwise you will encounter the `circular import error`. One way to fix this is to rename python file or function. Or you can use lazy-import (import inside a function), in this case you can import inside W4AFp8TPMoEMethod.apply()

**Comment by yuhyao** [comment]
@chenxijun1029 Nice work! Just wondering if there will be any further updates?
Also, should the file `sgl-kernel/tests/test_cutlass_w4a8_moe_mm.py` be updated as well?

**Comment by zhilingjiang** [comment]
Nice work!

**Comment by chenxijun1029** [comment]
> Thanks for your great work on this! To help us evaluate the impact of this PR, could you please provide the performance results (like GSM8K, MMLU, and Hellaswag)

Hi, I have updated the accuracy in the pr, please check it.

File: sgl-kernel/csrc/moe/cutlass_moe/w4a8/w4a8_grouped_mm_c3x.cu [hunk 5]
```diff
@@ -189,21 +192,164 @@ void dispatch_w4a8_moe_mm_sm90(
           s_strides,
           chunk_size);
     }
+  } else if (n == 512 && k == 7168) {
```
Comment (AniZpZ) [inline sgl-kernel/csrc/moe/cutlass_moe/w4a8/w4a8_grouped_mm_c3x.cu:205]: I think it's better to use a template to reduce duplicate code here
```
template<int TileM, int TileN, int TileK, int Stages, ...>
void invoke_gemm(...) {
    ...
    cutlass_w4a8_group_gemm_caller<GemmKernel>(...);
}
```

**Review by AniZpZ — APPROVED** [review]

**Comment by AniZpZ** [comment]
LGTM, do you have any more sugguestion? @yangsijia-serena

**Comment by yangsijia-celina** [comment]
> LGTM, do you have any more sugguestion? @yangsijia-serena

LGTM!

File: python/sglang/srt/layers/moe/fused_moe_triton/layer.py [hunk 6]
```diff
@@ -593,8 +595,9 @@ def _weight_loader_impl(
 
             if (
                 "compressed" in self.quant_method.__class__.__name__.lower()
-                and param.data[expert_id] != 1
-                and (param.data[expert_id] - loaded_weight).abs() > 1e-5
+                or "w4afp8" in self.quant_config.get_name()
```
Comment (BBuf) [inline python/sglang/srt/layers/moe/fused_moe_triton/layer.py:598]: It broken compressed-tensor format moe models like `neuralmagic/Mixtral-8x7B-Instruct-v0.1-FP8 ` , fixed in https://github.com/sgl-project/sglang/pull/10299 .

**Comment by Bruce-x-1997** [comment]
hello, thanks for the pr, but when I use w4afp8 in deepseek-r1(https://huggingface.co/Barrrrry/DeepSeek-R1-W4AFP8) and test it in aime24, it could only get 10% score
so what's your test model, which reach 80% in aime24 
@chenxijun1029

**Comment by chenxijun1029** [comment]
> hello, thanks for the pr, but when I use w4afp8 in deepseek-r1(https://huggingface.co/Barrrrry/DeepSeek-R1-W4AFP8) and test it in aime24, it could only get 10% score so what's your test model, which reach 80% in aime24 @chenxijun1029

I tested it with evalscope, and you should set "max_num_tokens" larger, i.e. 20000, for better performance.

**Comment by Bruce-x-1997** [comment]
> > hello, thanks for the pr, but when I use w4afp8 in deepseek-r1(https://huggingface.co/Barrrrry/DeepSeek-R1-W4AFP8) and test it in aime24, it could only get 10% score so what's your test model, which reach 80% in aime24 @chenxijun1029
> 
> I tested it with evalscope, and you should set "max_num_tokens" larger, i.e. 20000, for better performance.

thanks, I have reached 80% in r1 & r1-0528(quant use modelopt)
but I found if I use the same method to v3.1 , v3.1 accuracy could not be accepted, in v3.1(10% in aime24), and if we use a question in aime24 to test v3.1(after quant, and batch is 1), its answer is wrong.
do you find similar problems?is there anything we can try? @chenxijun1029

**Comment by chenxijun1029** [comment]
> > > hello, thanks for the pr, but when I use w4afp8 in deepseek-r1(https://huggingface.co/Barrrrry/DeepSeek-R1-W4AFP8) and test it in aime24, it could only get 10% score so what's your test model, which reach 80% in aime24 @chenxijun1029
> > 
> > 
> > I tested it with evalscope, and you should set "max_num_tokens" larger, i.e. 20000, for better performance.
> 
> thanks, I have reached 80% in r1 & r1-0528(quant use modelopt) but I found if I use the same method to v3.1 , v3.1 accuracy could not be accepted, in v3.1(10% in aime24), and if we use a question in aime24 to test v3.1(after quant, and batch is 1), its answer is wrong. do you find similar problems?is there anything we can try? @chenxijun1029

Can we share more information via wechat?

**Comment by llc-kc** [comment]
@chenxijun1029 do we need an extra launch parameter --quantization w4afp8?
